lib/CodeGen/MemoryAllocator.cpp

```diff
@@ -138,49 +154,573 @@ uint64_t MemoryAllocator::allocate(uint64_t size, Handle handle,
+    // end of all the previously allocated segments which were previously
+    // found. Since the previous segments are ordered by their stop address
+    // in ascending order this procedure is guaranteed to find a place at
+    // least at the end of the last segment.
+    // -------------------------------------------------------------------------
+
+    uint64_t currSegAddrStart = 0;
+    uint64_t currSegAddrStop = 0;
+
+    for (size_t prevSegIdx = 0; prevSegIdx < prevSegAddr.size(); prevSegIdx++) {
+
+      // Try to place current segment after this previously allocated segment.
+      currSegAddrStart = prevSegAddr[prevSegIdx].second;
+      currSegAddrStop = currSegAddrStart + buffSizeArray[currSegId];
+
+      // Verify if this placement overlaps with all the other segments.
+      // Note that this verification with all the previous segments is required
+      // because the previous segments can overlap between themselves.
+      bool overlap = false;
+      for (size_t ovrSegIdx = 0; ovrSegIdx < prevSegAddr.size(); ovrSegIdx++) {
+        // Check overlap.
+        overlap = overlap || intervalsOverlap(currSegAddrStart, currSegAddrStop,
+                                              prevSegAddr[ovrSegIdx].first,
+                                              prevSegAddr[ovrSegIdx].second);
+        // Early break if overlaps.
+        if (overlap) {
+          break;
+        }
+      }
+
+      // If no overlap than we found the solution for the placement.
+      if (!overlap) {
+        break;
+      }
+    }
+
+    // Update maximum used size.
+    usedSizeMax = std::max(usedSizeMax, currSegAddrStop);
+
+    // If max available memory is surpassed with the new segment then we stop
+    // the allocation and return early.
+    if (memorySize && (usedSizeMax > memorySize)) {
+      return MemoryAllocator::npos;
+    }
+
+    // Allocate current segment.
+    Segment currSeg(currSegAddrStart, currSegAddrStop);
+    idSegMap.insert(std::make_pair(currSegId, currSeg));
+
+    // Update buffer liveness information.
+    for (size_t allocIdx = buffTimeStartArray[currSegId];
+         allocIdx < buffTimeStopArray[currSegId]; allocIdx++) {
+      // Update total live sizes.
+      liveBuffSizeArray[allocIdx] -= buffSizeArray[currSegId];
+      // Update total live IDs.
+      auto &allocIds = liveBuffIdListArray[allocIdx];
+      auto it = std::find(allocIds.begin(), allocIds.end(), currSegId);
+      assert(it != allocIds.end() && "Buffer ID not found for removal!");
+      allocIds.erase(it);
+    }
+  }
+
+  // Verify again that all the buffers were allocated.
+  for (size_t allocIdx = 0; allocIdx < allocNum; allocIdx++) {
+    assert(liveBuffSizeArray[allocIdx] == 0 &&
+           "Not all buffers were allocated!");
+    assert(liveBuffIdListArray[allocIdx].empty() &&
+           "Not all buffers were allocated!");
+  }
+
+  return usedSizeMax;
+}
+
+uint64_t MemoryAllocator::allocateAll(const std::list<Allocation> &allocList) {
```

Comment:
This function is rather complex. Maybe split it into a main function and a bunch of helpers?

Reply:
Ok, I`ll look into this.

Reply:
I removed the part which maps handles to IDs and put it in a helper.
In order to make also the part which gathers buffer information as a separate utility I would need to make all the other definitions part of the MemoryAllocator class and that I think would pollute the header file with all sorts of stuff which are only specific to the implementation and not the API. WDYT?

Reply:
Hmm. 

> I think would pollute the header file with all sorts of stuff which are only specific to the implementation and not the API.

Could it help if you add something like this to the MemoryAllocator.h:

```c++
// Forward declaration of MemoryAllocatorImpl.
class MemoryAllocatorImpl.
...
class MemoryAllocator {
friend class MemoryAllocator;
...
};
```

So, basically you say there is friend class that can access all the info from MemoryAllocator. And then you define the actual implementation of a friend class MemoryAllocatorImpl in the cpp file.

WDYT? Would it work?